### PR TITLE
release: promote staging to main

### PR DESCRIPTION
**Pull request by harrymove-ctrl**
Routine promotion. 64 commits, 83 files, 15 PRs merged into `staging` since #796.

CI on the `staging` head is green, and every test suite passes locally (1064 tests, 0 failures — breakdown at the bottom).

---

## Before you merge: one thing to fix first

**Short version:** merging this publishes JS SDK `0.1.6` to npm, but leaves `main` saying `0.1.5`. After that, the JS SDK stops publishing entirely — silently, with CI still green.

**One-word fix**, in `.changeset/config.json`:

```diff
- "commit": true
+ "commit": false
```

Land that first, then merge this. Nothing else needs to change.

<details>
<summary>Why this happens (click to expand)</summary>

This PR carries `.changeset/recall-write-time-and-recency-weights.md`, which came in with #810. The file is correct — it just tells the release workflow "bump the JS SDK by one patch."

The workflow that reads it has a bug. `release-sdk.yml`:

```yaml
pnpm changeset version 2>/dev/null || true   # bumps 0.1.5 → 0.1.6 AND commits it
if git diff --quiet; then                    # tree is clean, because it was committed
  echo "has_changes=false"                   # → wrong conclusion: "nothing to push"
```

`.changeset/config.json` sets `"commit": true`, so `changeset version` commits the bump itself. `git diff --quiet` only looks at *uncommitted* changes, so it always sees a clean tree and reports `has_changes=false`. The `git push` step is then skipped.

But the publish step is gated only on `github.ref == 'refs/heads/main'`, so it runs anyway, reads the locally-bumped `package.json`, and publishes `0.1.6`.

| | repo | npm |
|---|---|---|
| now | 0.1.5 | 0.1.5 |
| after merging as-is | 0.1.5 | **0.1.6** |

The published `0.1.6` is fine — correct code, correct provenance. Nobody is broken. The damage is to the pipeline: on every later push to `main`, changesets recomputes `0.1.5` + the still-present changeset file → `0.1.6` again, `npm view …@0.1.6` now succeeds, and publish is skipped. **The JS SDK never publishes again**, and CI stays green the whole time.

This has been latent since March (`"commit": true` in 153e373f, the `git diff` probe in 71fcd0b7) because MCP and the Python SDK are version-bumped by hand in their feature branches — so `changeset version` has always been a no-op. This is the first time a changeset file reaches `main`.

`release-mcp.yml` has the same pattern and the same fix.

</details>

**Also:** merging triggers a mainnet Walrus Site deploy, because `apps/app/**` changed and that is the path filter on `deploy-app-walrus.yml`.

---

## What ships

| Area | Files | Highlights |
|---|:-:|---|
| `services/server` | 23 | `sort=recent` recall mode + `created_at` on results (#810) · `OWNER_TOKEN_SECRET` minimum-length check at startup (#835) · 400 instead of a provider error on oversized recall queries (#837) · clearer relayer job errors (#803) |
| `packages/mcp` | 18 | identify the coding agent on each session (#805) · sandbox credentials in login tests (#705) · plain-language login hand-off failures (#725) |
| `docs` | 13 | plugin-vs-MCP-only positioning + per-client install table (#848) · system prompt templates (#808) · MCP score vs SDK distance polarity (#849) · security-sweep reconcile (#844) |
| `.github` | 8 | SEAL cross-account deny invariant gated in CI (#841) |
| `packages/sdk` | 7 | `sort` and `created_at` on `recall()` |
| `packages/python-sdk-memwal` | 5 | `remember_bulk_async` rejects mismatched `job_ids` (#838) |

**Versions:** `memwal-mcp` 0.0.11 → 0.0.12 · `memwal` (Python) 0.1.8 → 0.1.9. Both hand-bumped, both publish correctly. The JS SDK is the only one going through changesets — see above.

### Two behaviour changes worth knowing

- **`OWNER_TOKEN_SECRET` shorter than 32 bytes now panics at startup.** Checked all three Railway environments before opening this: production and staging have it unset (feature disabled, boots normally), dev has 64 bytes. No environment is affected.
- **`rememberManual` / `remember_manual` now send `encryptedData` instead of a pre-uploaded `blobId`** in both SDKs. Breaking for anyone on the manual-embed path.

---

## Test status

| Suite | Result |
|---|---|
| `cargo test --bins` (server) | 670 passed, 0 failed, 37 ignored |
| `packages/mcp` | 60 passed, 0 failed |
| `packages/sdk` | 97 passed, 0 failed |
| `packages/python-sdk-memwal` | 146 passed, 16 skipped |
| `apps/app` (vitest) | 91 passed, 12 files |

The 37 ignored Rust tests need Postgres and Redis. Playwright e2e was not run.

**Not covered by any of the above:** `sort=recent` end-to-end against a live relayer — the 8 tests covering it are unit tests over the sort function and none reach `/api/recall` — and the MCP login/logout/concurrency fixes against a real client, since those 60 tests use a mock bridge. Worth a manual pass on dev after merging.

### CODE REVIEW

**Comment by jessiemongeon1** [comment]
## Style Guide Audit

Audited **17** file(s) against the [Sui Documentation Style Guide](https://docs.sui.io/references/contribute/style-guide).

**11** violation(s) found. All must be fixed before merge.

### `docs/mcp/overview.md` (1 violation(s))

_1 violation(s) (0 regex, 1 claude)_

- **Line 86** — quotation-marks
  - Current: `**"ChatGPT" means two different things here.**`
  - Fix: `**ChatGPT means two different things here.**`

### `docs/mcp/reference.md` (4 violation(s))

_4 violation(s) (1 regex, 3 claude)_

- **Line 78** — No em dashes in prose
  - Current: `—`
  - Fix: `Rewrite with comma, parentheses, or split sentence`
- **Line 78** — no-em-dash
  - Current: `— that is the cutoff biting, not an empty`
  - Fix: `. That is the cutoff biting, not an empty`
- **Line 85** — passive-voice
  - Current: `Hits with `distance >= maxDistance` are dropped.`
  - Fix: `The tool drops hits with `distance >= maxDistance`.`
- **Line 76** — passive-voice
  - Current: `hits with `distance >= maxDistance` are dropped`
  - Fix: `the tool drops hits with `distance >= maxDistance``

### `docs/python-sdk/changelog.mdx` (5 violation(s))

_5 violation(s) (0 regex, 5 claude)_

- **Line 45** — passive-voice
  - Current: `is reported as a retryable upstream outage, not a credential failure`
  - Fix: `the SDK reports it as a retryable upstream outage, not a credential failure`
- **Line 47** — passive-voice
  - Current: `restore `truncated` docs now match WALM-431 retryable semantics`
  - Fix: `Clarify subject: restore `truncated` docs now match WALM-431 retryable semantics (not passive, but line 47 is fine—disregard)`
- **Line 48** — passive-voice
  - Current: `invalid URLs are left for the HTTP client to surface`
  - Fix: `the SDK leaves invalid URLs for the HTTP client to surface`
- **Line 48** — passive-voice
  - Current: `so URL userinfo is not written to logs`
  - Fix: `so the SDK does not write URL userinfo to logs`
- **Line 45** — passive-voice
  - Current: `HTTP 503 from the relayer is reported as a retryable upstream outage`
  - Fix: `The SDK reports HTTP 503 from the relayer as a retryable upstream outage`

### `docs/relayer/synthetic-seal-cross-account.md` (1 violation(s))

_1 violation(s) (1 regex, 0 claude)_

- **Line 40** — Headings use sentence case (acronyms stay capitalized)
  - Current: `SEAL Cross-Account Synthetic`
  - Fix: `Capitalize only first word, proper nouns, and acronyms`

---
_Automated audit using the Sui Documentation Style Guide._

**Comment by ducnmm** [comment]
Confirming option 2 from the PR body: **no changesets**. Drop `.changeset/recall-write-time-and-recency-weights.md` and bump `@mysten-incubation/memwal` `0.1.5` → `0.1.6` by hand (`packages/sdk/package.json`, `packages/sdk/CHANGELOG.md`, `docs/sdk/changelog.mdx`, and the pin in `scripts/verify-manual-sdk-release.mjs`) — same path as MCP `0.0.12` and Python `0.1.9` in this PR.

Do not merge until that lands, or npm publishes `0.1.6` while `main` stays `0.1.5` as described. The mainnet Walrus Site deploy on merge is still expected once the version dump is in.

Leaving the PR open; not merging from here.

**Comment by ducnmm** [comment]
Style-guide nits from the audit plus the hand bump (drop changeset, JS SDK 0.1.6) are on https://github.com/MystenLabs/MemWal/pull/864 into `staging`. Once that merges, this promotion PR picks them up. Still not merging #861 from here.

**Comment by ducnmm** [comment]
Update after #865 merged to `staging`:

- Leftover changeset is gone. JS SDK is `0.1.6` in-tree (`packages/sdk/package.json` + changelogs + verify-script pin). Merge to `main` should publish `0.1.6` with git matching npm.
- The seven style-guide nits from the 04:20 audit are fixed on this head (`c3c1941e`). That comment is stale; a re-run should be clean.
- CI is green. Still needs a review approval. Merging still deploys the Walrus Site to mainnet (`apps/app/**`).

**Comment by ducnmm** [comment]
Please wait until the demo-app PRs merge to `dev` before noon ICT today (2026-09-07), then merge this staging → main / prod promotion before 17:00 ICT today.

Do not merge #861 before those PRs land (and are promoted through staging as needed), so prod does not miss the chatbot / noter / researcher fixes.

File: services/server/src/services/embedder.rs [hunk 1]
```diff
@@ -92,16 +105,18 @@ impl Embedder for OpenAiEmbedder {
...
                     if crate::services::extractor::is_upstream_status_transient(status) {
                         return Err(AppError::UpstreamUnavailable(format!(
                             "Embedding API upstream error ({}): {}",
                             status, body
                         )));
                     }
+                    if status == reqwest::StatusCode::BAD_REQUEST {
+                        tracing::warn!(%status, body, "embedding API returned 400");
+                        return Err(AppError::BadRequest(
+                            "embedding input exceeds the model context limit".into(),
```
Comment (ducnmm) [inline services/server/src/services/embedder.rs:117]: **[suggestion]** After the 16 KiB local gate, every non-transient HTTP 400 from the embedding API is rewritten as `BadRequest("embedding input exceeds the model context limit")`. OpenRouter/OpenAI 400s also cover invalid model, malformed body, and similar config errors. Callers then retry by shrinking text while operators only have the `tracing::warn` body.

**Suggestion:** Map to a generic 400 (or inspect the upstream body for a context-length signal) instead of always blaming the model limit.

File: services/server/src/types.rs [hunk 2]
```diff
@@ -1363,6 +1452,26 @@ pub struct RecallResult {
...
+    ///
+    /// Present so a caller can order and verify the returned set by
+    /// write-time rather than re-ranking on a date it has to parse back out
+    /// of the memory text (WALM-383). Note this is the *write* time, not any
+    /// event time the text itself may describe.
+    ///
+    /// This alone does not make "newest wins" correct: the candidate set is
+    /// the cosine top-`limit` from `search_similar`, so the newest row can be
+    /// missing from `results` entirely and no client-side sort recovers it.
+    /// The server-side recency mode that over-fetches is still open.
```
Comment (ducnmm) [inline services/server/src/types.rs:1494]: **[nit]** `RecallResult.created_at` still says the server-side over-fetch recency mode “is still open”. This PR adds `RecallSort::Recent` and `candidate_limit` in the same file, so the comment is already false.

**Suggestion:** Point at `RecallSort::Recent` (default recall is still cosine top-`limit`).

**Review by ducnmm — APPROVED** [review]

**Comment by ducnmm** [comment]
@harrymove-ctrl this is yours to merge (you authored it; same as #874). Daily will not merge.

Leftover changeset is gone on this head (`32c22759`). TS 0.1.6 / MCP 0.0.12 / Python 0.1.9 will publish on merge.

Please confirm Railway **prod** `AUTH_SECRET` is ≥32 chars (and not a weak-but-long value) before merging — #834 500s researcher otherwise. Staging already rotated.
